geiger/src/graph/Graph.js

```diff
@@ -0,0 +1,283 @@
+/**
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Collection from '../util/Collection';
+import Events from '../util/Events';
+import NodeWithPort from './NodeWithPort';
+import Edge, {generateEdgePortIdsByMessage} from './Edge';
+import {EdgeTypes} from './EdgeTypes';
+
+// /<reference path="../jsdoc.types.js" />
+
+/**
+ * A module that tracks all the nodes and edges.
+ *
+ * @class Graph
+ * @extends Events
+ */
+export default class Graph extends Events {
+  /**
+   * @constructor
+   * @param {!string} contextId
+   */
+  constructor(contextId) {
+    super();
+
+    this.contextId = contextId;
+
+    this._nodes = new Collection(null, {trackExit: true});
+    this._edges = new Collection(null, {trackExit: true});
+
+    // Cache the removed nodes/edges in current batch.
+    this._removed_nodes = [];
+    this._removed_edges = [];
+
+    /**
+     * For each node ID, keep track of all out-bound edge IDs.
+     * @type {!Object<string, !Object<string, true>>}
+     * @private
+     */
+    this._outbound_edge_map = {};
+
+    /**
+     * For each node ID, keep track of all in-bound edge IDs.
+     * @type {!Object<string, !Object<string, true>>}
+     * @private
+     */
+    this._inbound_edge_map = {};
+  }
+
+  /**
+   * Add a node to the graph.
+   * @param {!NodeCreationMessage} message
+   */
+  addNode(message) {
+    const node = new NodeWithPort(message);
+    this._nodes.add(message.nodeId, node);
+    this._shouldRedraw();
+  }
+
+  /**
+   * Remove a node by id and all related edges.
+   * @param {!string} nodeId
+   */
+  removeNode(nodeId) {
+    this._nodes.remove(nodeId);
+    this._removeEdgesOfNode(nodeId);
+    this._shouldRedraw();
+  }
+
+  /**
+   * Add a param to the node.
+   * @param {!ParamCreationMessage} message
```

Comment:
`Message` feels like a bunch of string needs to be printed. How about `ParamCreationData`?

Reply:
Done.